File: dummy_client.cpp
#include <iostream>
#include <string>
#include <cstring>
#include <cstdlib>
#include <ctime>
#include "SDL2/SDL.h"
#include "SDL2/SDL_net.h"

// Network constants (matching the main game)
const int PLAYER_STATE_CMD = 6;

class DummyClient {
private:
    TCPsocket socket;
    std::string playerName;
    int x, y;
    int direction;
    bool connected;
    
public:
    DummyClient(const std::string& name, int startX, int startY) 
        : playerName(name), x(startX), y(startY), direction(0), connected(false), socket(nullptr) {
    }
    
    ~DummyClient() {
        if (socket) {
            SDLNet_TCP_Close(socket);
        }
    }
    
    bool connect() {
        IPaddress ip;
        if (SDLNet_ResolveHost(&ip, "localhost", 8099) == -1) {
            std::cerr << "SDLNet_ResolveHost error: " << SDLNet_GetError() << std::endl;
            return false;
        }
        
        socket = SDLNet_TCP_Open(&ip);
        if (socket == NULL) {
            std::cerr << "SDLNet_TCP_Open error: " << SDLNet_GetError() << std::endl;
            return false;
        }
        
        connected = true;
        std::cout << "Dummy client '" << playerName << "' connected to server" << std::endl;
        return true;
    }
    
    void sendPlayerState() {
        if (!connected || !socket) return;
        
        unsigned char packet[64];
        int offset = 0;
        
        // [command(4)][id(4)][x(4)][y(4)][direction(4)][nameLen(1)][name]
        int cmd = PLAYER_STATE_CMD;
        int id = 0; // Server will assign proper ID
        
        memcpy(packet + offset, &cmd, 4); offset += 4;
        memcpy(packet + offset, &id, 4); offset += 4;
        memcpy(packet + offset, &x, 4); offset += 4;
        memcpy(packet + offset, &y, 4); offset += 4;
        memcpy(packet + offset, &direction, 4); offset += 4;
        
        unsigned char nameLen = (unsigned char)std::min((int)playerName.length(), 32);
        packet[offset++] = nameLen;
        memcpy(packet + offset, playerName.c_str(), nameLen);
        
        int total = offset + nameLen;
        int bytesSent = SDLNet_TCP_Send(socket, packet, total);
        
        if (bytesSent < total) {
            std::cerr << "Failed to send player state for " << playerName << std::endl;
            connected = false;
        } else {
            std::cout << playerName << " sent position (" << x << ", " << y << ")" << std::endl;
        }
    }
    
    void simulateMovement() {
        // Simple random movement pattern
        int moveX = (rand() % 3 - 1) * 12; // -12, 0, or 12 pixels
        int moveY = (rand() % 3 - 1) * 12;
        
        x += moveX;
        y += moveY;
        
        // Keep within reasonable bounds
        if (x < 0) x = 0;
        if (y < 0) y = 0;
        if (x > 720) x = 720;
        if (y > 480) y = 480;
        
        // Update direction based on movement
        if (moveX > 0) direction = 2; // East
        else if (moveX < 0) direction = 3; // West
        else if (moveY > 0) direction = 0; // South
        else if (moveY < 0) direction = 1; // North
    }
    
    bool isConnected() const { return connected; }
};

int main(int argc, char* argv[]) {
    // Parse command line arguments
    std::string playerName = "DummyPlayer";
    int startX = 100;
    int startY = 100;
    int moveInterval = 2000; // milliseconds between moves
    
    if (argc >= 2) playerName = argv[1];
    if (argc >= 3) startX = std::atoi(argv[2]);
    if (argc >= 4) startY = std::atoi(argv[3]);
    if (argc >= 5) moveInterval = std::atoi(argv[4]);
    
    std::cout << "Starting dummy client: " << playerName << " at (" << startX << ", " << startY << ")" << std::endl;
    
    // Initialize SDL networking
    if (SDL_Init(0) == -1) {
        std::cerr << "SDL_Init failed: " << SDL_GetError() << std::endl;
        return 1;
    }
    
    if (SDLNet_Init() == -1) {
        std::cerr << "SDLNet_Init failed: " << SDLNet_GetError() << std::endl;
        return 2;
    }
    
    srand(time(nullptr));
    
    // Create and connect dummy client
    DummyClient client(playerName, startX, startY);
    
    if (!client.connect()) {
        std::cerr << "Failed to connect to server" << std::endl;
        SDLNet_Quit();
        SDL_Quit();
        return 3;
    }
    
    // Send initial position
    client.sendPlayerState();
    
    // Main loop - send position updates periodically
    Uint32 lastMoveTime = SDL_GetTicks();
    
    std::cout << "Dummy client running. Press Ctrl+C to stop." << std::endl;
    
    while (client.isConnected()) {
        Uint32 currentTime = SDL_GetTicks();
        
        if (currentTime - lastMoveTime >= moveInterval) {
            client.simulateMovement();
            client.sendPlayerState();
            lastMoveTime = currentTime;
        }
        
        SDL_Delay(100); // Small delay to avoid busy waiting
    }
    
    std::cout << "Dummy client " << playerName << " disconnected" << std::endl;
    
    SDLNet_Quit();
    SDL_Quit();
    return 0;
}

File: dummy_client.py
#!/usr/bin/env python3
"""
Simple Python dummy client for testing Ultimata multiplayer
Usage: python dummy_client.py [name] [x] [y] [host] [port]
"""

import socket
import struct
import time
import random
import sys
import threading

class DummyClient:
    def __init__(self, name="DummyPlayer", start_x=100, start_y=100, host="localhost", port=8099):
        self.name = name[:15]  # Max 15 chars for name
        self.x = float(start_x)
        self.y = float(start_y) 
        self.direction = 0
        self.player_id = random.randint(1000, 9999)  # Random ID
        self.host = host
        self.port = port
        self.sock = None
        self.connected = False
        self.running = True
        
    def connect(self):
        """Connect to the game server"""
        try:
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.sock.connect((self.host, self.port))
            self.connected = True
            print(f"[{self.name}] Connected to {self.host}:{self.port}")
            return True
        except Exception as e:
            print(f"[{self.name}] Connection failed: {e}")
            return False
    
    def disconnect(self):
        """Disconnect from server"""
        if self.sock:
            try:
                # Send disconnect message
                disconnect_msg = "disconnect\n"
                self.sock.send(disconnect_msg.encode())
            except:
                pass
            self.sock.close()
        self.connected = False
        self.running = False
        print(f"[{self.name}] Disconnected")
    
    def send_player_state(self):
        """Send binary PLAYER_STATE packet to server"""
        if not self.connected:
            return False
            
        try:
            # Create binary packet: [command][id][x][y][direction][nameLen][name]
            # PLAYER_STATE command = 6
            command = 6
            name_bytes = self.name.encode('utf-8')
            name_len = len(name_bytes)
            
            # Pack as binary data (little-endian)
            packet = struct.pack('<I', command)  # command (4 bytes)
            packet += struct.pack('<I', self.player_id)  # id (4 bytes)  
            packet += struct.pack('<I', int(self.x))  # x (4 bytes)
            packet += struct.pack('<I', int(self.y))  # y (4 bytes)
            packet += struct.pack('<I', self.direction)  # direction (4 bytes)
            packet += struct.pack('<B', name_len)  # nameLen (1 byte)
            packet += name_bytes  # name (nameLen bytes)
            
            self.sock.send(packet)
            return True
        except Exception as e:
            print(f"[{self.name}] Failed to send state: {e}")
            return False
    
    def simulate_movement(self):
        """Simulate random movement"""
        # Random walk with bounds
        dx = random.uniform(-2.0, 2.0)
        dy = random.uniform(-2.0, 2.0)
        
        self.x += dx
        self.y += dy
        
        # Keep within reasonable bounds
        self.x = max(50, min(800, self.x))
        self.y = max(50, min(600, self.y))
        
        # Random direction change occasionally
        if random.random() < 0.1:
            self.direction = random.randint(0, 3)
    
    def run(self):
        """Main client loop"""
        if not self.connect():
            return
            
        print(f"[{self.name}] Starting at ({int(self.x)}, {int(self.y)})")
        print(f"[{self.name}] Player ID: {self.player_id}")
        print(f"[{self.name}] Press Ctrl+C to stop")
        
        try:
            while self.running:
                # Send current state
                if not self.send_player_state():
                    break
                    
                # Simulate movement
                self.simulate_movement()
                
                # Print position occasionally
                if random.random() < 0.05:  # 5% chance
                    print(f"[{self.name}] Position: ({int(self.x)}, {int(self.y)})")
                
                # Update every 100ms
                time.sleep(0.1)
                
        except KeyboardInterrupt:
            print(f"\n[{self.name}] Stopping...")
        except Exception as e:
            print(f"[{self.name}] Error: {e}")
        finally:
            self.disconnect()

def main():
    # Parse command line arguments
    name = "Player" + str(random.randint(1, 999))
    x = 100 + random.randint(0, 300)
    y = 100 + random.randint(0, 200) 
    host = "localhost"
    port = 8099
    
    if len(sys.argv) > 1:
        name = sys.argv[1]
    if len(sys.argv) > 2:
        x = int(sys.argv[2])
    if len(sys.argv) > 3:
        y = int(sys.argv[3])  
    if len(sys.argv) > 4:
        host = sys.argv[4]
    if len(sys.argv) > 5:
        port = int(sys.argv[5])
    
    print("=== Ultimata Dummy Client ===")
    print(f"Name: {name}")
    print(f"Starting position: ({x}, {y})")
    print(f"Server: {host}:{port}")
    print()
    
    client = DummyClient(name, x, y, host, port)
    client.run()

if __name__ == "__main__":
    main()